python/ray/serve/_private/proxy.py

```diff
@@ -569,68 +568,46 @@ async def not_found(self, proxy_request: ProxyRequest) -> ProxyResponse:
             f"{application_message} Please ping "
             "/ray.serve.RayServeAPIService/ListApplications for available applications."
         )
-        status_code = grpc.StatusCode.NOT_FOUND
-        proxy_request.send_status_code(status_code=status_code)
-        proxy_request.send_details(message=not_found_message)
-        return ProxyResponse(status_code=str(status_code))
-
-    async def draining_response(self, proxy_request: ProxyRequest) -> ProxyResponse:
-        status_code = grpc.StatusCode.UNAVAILABLE
-        proxy_request.send_status_code(status_code=status_code)
-        proxy_request.send_details(message=DRAINED_MESSAGE)
+
+        yield ResponseStatus(
+            code=grpc.StatusCode.NOT_FOUND,
+            message=not_found_message,
+            is_error=True,
+        )
+
+    async def draining_response(self, proxy_request: ProxyRequest) -> ResponseGenerator:
         if proxy_request.is_route_request:
             application_names = [endpoint.app for endpoint in self.route_info.values()]
             response_proto = ListApplicationsResponse(
                 application_names=application_names
             )
         else:
             response_proto = HealthzResponse(message=DRAINED_MESSAGE)
-        return ProxyResponse(
-            status_code=str(status_code),
-            response=response_proto.SerializeToString(),
-        )
 
-    async def timeout_response(
-        self, proxy_request: ProxyRequest, request_id: str
-    ) -> ProxyResponse:
-        timeout_message = (
-            f"Request {request_id} timed out after {self.request_timeout_s}s."
+        yield response_proto.SerializeToString()
```

Comment:
For making my review easier :) , why do we need to have the first yield?

Reply:
Yes, this will return the message to the client first. Then the second yield `ResponseStatus` is to used internally for logging :)

Reply:
sync offline, understood